### ISSUE TITLE
vsphere hive deployments broken as of v1.0.16

### ISSUE THREAD

**Issue by rayashworth** [openshift/hive#1235]
been test vsphere hive deployments and it would appear that v1.0.16 has a breaking change.

Scenarios:
- Install OCP cluster using ACM UI hive version v1.0.14. works fine
- install OCP cluster using direct call to hive (v1.0.14) instance under ACM, works fine.
- install OCP cluster using hive v1.0.14 public operator. Works fine.
- install OCP clsuter using hive v1.0.16, not working.
- install OCP cluster using hive v1.0.15, not working. does not appear to create all of the hive pods after deploy.

error messages (snippet)
time="2020-12-17T14:49:58.259Z" level=info msg="reconcile complete" controller=metrics elapsedMillis=0 elapsedMillisGT=0 outcome=unspecified
time="2020-12-17T14:51:58.259Z" level=info msg="calculating metrics across all ClusterDeployments" controller=metrics
time="2020-12-17T14:51:58.259Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-17T14:51:58.26Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-17T14:51:58.26Z" level=debug msg="calculating metrics across all install jobs" controller=metrics
time="2020-12-17T14:51:58.26Z" level=debug msg="calculating metrics across all uninstall jobs" controller=metrics
time="2020-12-17T14:51:58.26Z" level=debug msg="calculating metrics across all imageset jobs" controller=metrics
time="2020-12-17T14:51:58.26Z" level=debug msg="calculating metrics across all ClusterSyncs" controller=metrics
time="2020-12-17T14:51:58.26Z" level=info msg="reconcile complete" controller=metrics elapsedMillis=0 elapsedMillisGT=0 outcome=unspecified
time="2020-12-17T14:53:58.26Z" level=info msg="calculating metrics across all ClusterDeployments" controller=metrics
time="2020-12-17T14:53:58.26Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-17T14:53:58.261Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-17T14:53:58.261Z" level=debug msg="calculating metrics across all install jobs" controller=metrics
time="2020-12-17T14:53:58.261Z" level=debug msg="calculating metrics across all uninstall jobs" controller=metrics
time="2020-12-17T14:53:58.261Z" level=debug msg="calculating metrics across all imageset jobs" controller=metrics
time="2020-12-17T14:53:58.261Z" level=debug msg="calculating metrics across all ClusterSyncs" controller=metrics
time="2020-12-17T14:53:58.261Z" level=info msg="reconcile complete" controller=metrics elapsedMillis=0 elapsedMillisGT=0 outcome=unspecified
time="2020-12-17T14:55:58.261Z" level=info msg="calculating metrics across all ClusterDeployments" controller=metrics
time="2020-12-17T14:55:58.261Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-17T14:55:58.261Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-17T14:55:58.262Z" level=debug msg="calculating metrics across all install jobs" controller=metrics
time="2020-12-17T14:55:58.262Z" level=debug msg="calculating metrics across all uninstall jobs" controller=metrics
time="2020-12-17T14:55:58.262Z" level=debug msg="calculating metrics across all imageset jobs" controller=metrics
time="2020-12-17T14:55:58.262Z" level=debug msg="calculating metrics across all ClusterSyncs" controller=metrics
time="2020-12-17T14:55:58.262Z" level=info msg="reconcile complete" controller=metrics elapsedMillis=0 elapsedMillisGT=0 outcome=unspecified
time="2020-12-17T14:57:52.33Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=289xsp46
time="2020-12-17T14:57:52.353Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=289xsp46
time="2020-12-17T14:57:52.353Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=289xsp46
time="2020-12-17T14:57:52.353Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=23 elapsedMillisGT=0 outcome=unspecified reconcileID=289xsp46

**Comment by dgoodwin** [comment]
@joelddiaz @abutcher any thoughts here?

Sounds like same cluster, standalone hive, exact same clusterdeployment definition (is this true?), 1.0.14 works, 1.0.16 breaks.

The only relevant change I see initially is the vmware cred checking in 58fe555836cb5c2ff618dbd1d18a8404ab6393b4.

**Comment by rayashworth** [comment]
yes exact same clusterdeployment

**Comment by rayashworth** [comment]
[vSpherex509Error.log](https://github.com/openshift/hive/files/5710878/vSpherex509Error.log)

**Comment by rayashworth** [comment]
v1.0.15 did not work at all, so I think we ignore that test.

**Comment by joelddiaz** [comment]
looks like the creds checking is failing:
```time="2020-12-17T14:57:52.353Z" level=warning msg="failed to validate VSphere credentials" ```

**Comment by rayashworth** [comment]
@dgoodwin @joelddiaz @abutcher Let me know if I can do any other testing that may help this along.

**Comment by joelddiaz** [comment]
I looked at this for a bit, and here's what I've found (just via code inspection as I don't have an untrusted vSphere endpoint to test against).

The vSphere platform allows providing some custom CAs at https://github.com/openshift/hive/blob/master/pkg/apis/hive/v1/vsphere/platform.go#L16-L18

When this is set, the hive provision/deprovision Pods are set up to add these CAs to the Pod's root CA trust list: https://github.com/openshift/hive/blob/master/pkg/install/generate.go#L750-L751 and https://github.com/openshift/hive/blob/master/pkg/install/generate.go#L342 .

When hive tries to perform the pre-provision validation of the vSphere connection, it does not use this extra CA information: https://github.com/openshift/hive/blob/master/pkg/controller/utils/credentials.go#L27-L36 .

I believe what we want to do is copy (or extend) the installer's vSphere connection test https://github.com/openshift/hive/blob/master/vendor/github.com/openshift/installer/pkg/types/vsphere/client.go#L14-L17 (which is what hive is using for validation) to allow the use of the SetRootCAs() call that the vmware client libraries expose https://github.com/openshift/hive/blob/master/vendor/github.com/vmware/govmomi/vim25/soap/client.go#L218-L223 .

Since the hive credentials checking runs in the hive controller Pod's context, we do not want to set/modify the Pod's root CAs trust list, and this SetRootCAs() should allow us to temporarily provide some certificates that are specific just to the specific vSphere ClusterDeployment that is being validated.

**Comment by rayashworth** [comment]
@joelddiaz Test your branch

time="2020-12-22T17:55:07.546Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=2fkkls49
time="2020-12-22T17:55:07.546Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=13 elapsedMillisGT=0 outcome=unspecified reconcileID=2fkkls49
time="2020-12-22T17:55:07.866Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=48dtwvrm
time="2020-12-22T17:55:07.881Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=48dtwvrm
time="2020-12-22T17:55:07.881Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=48dtwvrm
time="2020-12-22T17:55:07.881Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=15 elapsedMillisGT=0 outcome=unspecified reconcileID=48dtwvrm
time="2020-12-22T17:55:08.522Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=47mh7mll
time="2020-12-22T17:55:08.534Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=47mh7mll
time="2020-12-22T17:55:08.534Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=47mh7mll
time="2020-12-22T17:55:08.534Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=12 elapsedMillisGT=0 outcome=unspecified reconcileID=47mh7mll
time="2020-12-22T17:55:09.816Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=4lc5lg4p
time="2020-12-22T17:55:09.829Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=4lc5lg4p
time="2020-12-22T17:55:09.829Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=4lc5lg4p
time="2020-12-22T17:55:09.829Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=13 elapsedMillisGT=0 outcome=unspecified reconcileID=4lc5lg4p
time="2020-12-22T17:55:12.389Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=fhh56vpx
time="2020-12-22T17:55:12.4Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=fhh56vpx
time="2020-12-22T17:55:12.4Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=fhh56vpx
time="2020-12-22T17:55:12.4Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=10 elapsedMillisGT=0 outcome=unspecified reconcileID=fhh56vpx
time="2020-12-22T17:55:17.52Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=27hfx899
time="2020-12-22T17:55:17.537Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=27hfx899
time="2020-12-22T17:55:17.537Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=27hfx899
time="2020-12-22T17:55:17.537Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=16 elapsedMillisGT=0 outcome=unspecified reconcileID=27hfx899
time="2020-12-22T17:55:27.777Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=k587dlds
time="2020-12-22T17:55:27.791Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=k587dlds
time="2020-12-22T17:55:27.791Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=k587dlds
time="2020-12-22T17:55:27.791Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=13 elapsedMillisGT=0 outcome=unspecified reconcileID=k587dlds
time="2020-12-22T17:55:41.825Z" level=info msg="calculating metrics across all ClusterDeployments" controller=metrics
time="2020-12-22T17:55:41.825Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-22T17:55:41.825Z" level=warning msg="condition type AuthenticationFailure missing from AllClusterDeploymentConditions slice"
time="2020-12-22T17:55:41.825Z" level=debug msg="calculating metrics across all install jobs" controller=metrics
time="2020-12-22T17:55:41.825Z" level=debug msg="calculating metrics across all uninstall jobs" controller=metrics
time="2020-12-22T17:55:41.825Z" level=debug msg="calculating metrics across all imageset jobs" controller=metrics
time="2020-12-22T17:55:41.825Z" level=debug msg="calculating metrics across all ClusterSyncs" controller=metrics
time="2020-12-22T17:55:41.825Z" level=info msg="reconcile complete" controller=metrics elapsedMillis=0 elapsedMillisGT=0 outcome=unspecified
time="2020-12-22T17:55:48.271Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=xmkslms9
time="2020-12-22T17:55:48.289Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=xmkslms9
time="2020-12-22T17:55:48.289Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=xmkslms9
time="2020-12-22T17:55:48.289Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=17 elapsedMillisGT=0 outcome=unspecified reconcileID=xmkslms9
time="2020-12-22T17:56:29.25Z" level=info msg="reconciling cluster deployment" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment reconcileID=p8gr795r
time="2020-12-22T17:56:29.268Z" level=warning msg="failed to validate VSphere credentials" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Post \"https://ICOVCPC65.rtp.raleigh.ibm.com/sdk\": x509: certificate signed by unknown authority" reconcileID=p8gr795r
time="2020-12-22T17:56:29.268Z" level=error msg="cannot proceed with provision while platform credentials authentication is failing." clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment error="Platform credentials failed authentication check" reconcileID=p8gr795r
time="2020-12-22T17:56:29.268Z" level=info msg="reconcile complete" clusterDeployment=ash-test-vsphere/ash-test-vsphere controller=clusterDeployment elapsedMillis=18 elapsedMillisGT=0 outcome=unspecified reconcileID=p8gr795r

**Comment by chrisahl** [comment]
With ACM 2.2, I was testing out the new hive v1.0.16 using the exact same VMware credentials/cert/etc that I successfully used with a previous hive (v1.0.13), and it fails with 
```
clusterDeployment error="Post \"https://10.1.158.10/sdk\": x509: certificate signed by unknown authority" 
``` 
Has something changed in the last few versions to cause a VMware regression?

**Comment by joelddiaz** [comment]
Yes. There were some changes to perform some pre-flight creds validation that merged recently https://github.com/openshift/hive/pull/1186

This causes an issue when connecting to a vSphere endpoint where the certificates are not trusted. The fix for that has not yet merged and can be seen https://github.com/openshift/hive/pull/1238

**Comment by chrisahl** [comment]
@dgoodwin Any ETA on when this fix will go in?  ACM needs this fix and a new hive published for us to pickup (v1.0.17)

**Comment by chrisahl** [comment]
This was fixed a long time ago with v1.0.17
